src/sentry/monitors/processing_errors.py

```diff
@@ -36,3 +54,127 @@ class ProcessingErrorType(Enum):
     """This monitor environment is sending checkins too frequently"""
     ORGANIZATION_KILLSWITCH_ENABLED = 15
     """We have disabled checkin ingestion for this org. Contact support for details"""
+
+
+class CheckinValidationError(Exception):
+    def __init__(self, processing_errors: list[ProcessingError], monitor: Monitor | None = None):
+        # Monitor is optional, since we don't always have the monitor related to the checkin available
+        self.processing_errors = processing_errors
+        self.monitor = monitor
+
+
+class ProcessingErrorData(TypedDict):
+    type: str
+    data: dict[str, Any]
+
+
+@dataclasses.dataclass(frozen=True)
+class ProcessingError:
+    type: ProcessingErrorType
+    data: dict[str, Any] = dataclasses.field(default_factory=dict)
+
+    def to_dict(self) -> ProcessingErrorData:
+        return {
+            "type": self.type.name,
+            "data": self.data,
+        }
+
+    @classmethod
+    def from_dict(cls, processing_error_data: ProcessingErrorData) -> ProcessingError:
+        return cls(
+            ProcessingErrorType[processing_error_data["type"]],
+            processing_error_data["data"],
+        )
+
+
+class CheckinProcessingErrorData(TypedDict):
+    errors: list[ProcessingErrorData]
+    checkin: CheckinItemData
+
+
+@dataclasses.dataclass(frozen=True)
+class CheckinProcessingError:
+    errors: list[ProcessingError]
+    checkin: CheckinItem
+
+    def to_dict(self) -> CheckinProcessingErrorData:
+        return {
+            "errors": [error.to_dict() for error in self.errors],
+            "checkin": self.checkin.to_dict(),
+        }
+
+    @classmethod
+    def from_dict(cls, data: CheckinProcessingErrorData) -> CheckinProcessingError:
+        return cls(
+            errors=[ProcessingError.from_dict(error) for error in data["errors"]],
+            checkin=CheckinItem.from_dict(data["checkin"]),
+        )
+
+
+class CheckinProcessErrorsManager:
```

Comment:
Do we actually need the class? Doesn't seem to have any interesting state, just using it as a namespace?

Reply:
Yeah, mostly seemed useful to keep it as a namespace rather than have functions scattered around. Could be a module instead?

Reply:
yeah could be a module. I'm indifferent